[Blazor] Improvements to the interaction between SSR and interactive rendering

MackinnonBuck opened the pull request:
# Improvements to the interaction between SSR and interactive rendering

Makes the following improvements to how SSR and interactive rendering interact:
- [X] Detect and initialize interactive root components added via an SSR update
- [X] Cleanly synchronize SSR updates with existing interactive components
  - [X] Updated parameter values get supplied
  - [X] Components get disposed when removed from the DOM

Fixes #48763
Fixes #48764

MackinnonBuck committed:
Stream updates can remove interactive components

MackinnonBuck committed:
Merge remote-tracking branch 'origin/main' into mbuck/ssr-boundary-improvements

SteveSandersonMS commented on src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Streaming.cs:
```diff
@@ -99,11 +98,35 @@ private void SendBatchAsStreamingUpdate(in RenderBatch renderBatch, TextWriter w
+                    // It should be noted that a non-interactive parent component might still SSR the
+                    // interactive component again, at which point the browser will remove and
+                    // re-initialize the component with updated parameter values.
+                    continue;
```
I've been thinking through whether or not we should prerender interactive components more than once.

At first I thought it was debatable and could go either way, but eventually realised that **we don't want components to start becoming interactive until streaming SSR is completed**. Because otherwise, we'd be wastefully discarding any "loading" work that is happening during prerendering, and would be unable to initialize the interactive component with any `PersistentComponentState` that is obtained from loading.

If that's true, then I don't see any reason *not* to send streaming updates to the prerendered content for all components equally, whether or not they are interactive. The server is doing this work to load and render data for them anyway; presumably the app developer doesn't want to discard the output from that work.

The more subtle case is if we know this request is for a progressively-enhanced navigation (so there may already be interactive components on the page). We do have a special HTTP header telling us if it's a PE-enhanced nav, and in an ideal situation, the server would somehow know which rendermode boundaries can now be ignored (because there's already an interactive version of that component in the browser) and would just emit updated parameters for them. But unfortunately the server *doesn't* know which rendermode boundaries can be ignored - maybe we're rendering a different set of components now. The only thing that can resolve this correspondence is the client, which will use its DOM sync logic to establish whether the newly-received interactive markers line up with existing interactive components, and in that case, it should just take the new parameters and discard the prerendered content. Clearly there's some wastefulness in that but AFAIK it's the best that's logically possible, or maybe we think of further optimizations in the future whereby the PE-nav requests somehow also communicate info about what interactive components they already have (not for .NET 8).

What do you think? Does this sound right or am I missing something?

MackinnonBuck replied:
Thanks for the feedback, @SteveSandersonMS!

I think I agree with you that newly-SSR'd components shouldn't become active until streaming rendering completes. I also went back and forth on this, and my initial reasoning for eagerly-initializing interactive components was:
* It might inhibit the user experience to delay interactivity for components that might not receive further stream rendering updates anyway.
* If a fully initialized version of a component looks significantly different than its initial render, making the component interactive as soon as possible minimizes how much the component reverts back to its "initial" state when interactivity finally starts.

However, the first point is probably not a big deal because unless stream rendering takes ages, the user experience isn't likely to degrade massively by a component not being interactive for a few seconds. And the second point is probably not a concern considering the point you made about `PersistentComponentState`.

---

I also agree with your analysis about what would be the most optimal behavior regarding the interaction between SSR and interactive components. Part of my reasoning for the current approach was that it made the logic super straightforward when an SSR update conflicts with an interactive component (SSR always wins, dispose and re-initialize the interactive component). I had prototyped some approaches for a more sophisticated merging behavior where the DOM sync algorithm would correctly know to ignore existing interactive parts of the DOM, which it would correctly map to incoming component comments, but I wasn't happy with any of those approaches.

I did some more thinking about this today and started implementing an approach that I believe should work fairly well. Most of the ideas in my current approach still apply. It’s not ready to commit yet but here's a summary of what I'm thinking:

* If a component gets SSR'd, it always produces output in the response, regardless of whether it was rendered already.
* Before the DOM sync occurs (in the `beforeDocumentUpdated` callback), we:
  * Find component markers in the "incoming" HTML using existing logic (`discoverComponents`).
  * Find interactive component trees in the "destination" content being updated, and move them out of the DOM to separate document fragments
    * This is similar to what happens in the current implementation, minus the eager component disposal
    * This allows the DOM sync algorithm to work without special logic handling interactively managed DOM
    * _Note:_ We won't have to handle SSR updates that involve only a subtree of interactive DOM. When a component becomes interactive, prerendered content (including boundary markers) gets removed, and the current stream rendering logic already ignores incoming content that doesn't have matching boundary markers.
    * _Another note:_ I don't anticipate that removing/reinserting chunks of DOM like this will create any visual flickering or shifting because all changes to the DOM here happen synchronously.
  * For each interactive root component previously identified in the destination:
    * If it has a corresponding component comment in the incoming markup, clear any associated prerendered content (leaving the comment intact).
    * Otherwise, initiate asynchronous disposal of the component.
* After the DOM sync occurs (in the `afterDocumentUpdated` callback), we:
  * Move existing interactive components back into the DOM at the locations of their corresponding newly inserted component comments
    * Apply updated parameters from the comment's payload.
    * Remove the comment.
  * Initialize interactive components for each comment that didn't correlate to an existing interactive component.

---

To make this work, we'll probably have to serialize a `componentId` in each component comment, representing the ID of the SSR'd component. That way, interactive components can be mapped to SSR'd comments.

Also, the Blazor Server component descriptor protocol currently has a security restriction that forces root components to be added in the order that they get rendered. This isn't always possible with SSR because:
* We might deliberately try to utilize SSR'd components in a different order than they were rendered (for example, using SSR output to update component parameters might happen before newly SSR'd components become interactive).
* SSR'd components might get ignored for interactivity outright (for example, a component gets added and then removed from the page in two separate SSR updates before interactivity starts).

We should be able to get around this problem by allowing Blazor Server component comments to be used in the "wrong" order, but then requiring the client to provide the sequence number of each component along with the data-protected payload. From a security perspective, I think this is equivalent to what we have today. The only difference is that instead of the sequence numbers being implied by the order of each component descriptor in the serialized array, they're stated explicitly.

**TL;DR:** I agree with your comment and I'm currently working on updates 🙂

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/SymbolUtil.ts:
```diff
@@ -0,0 +1,6 @@
+
+export function createSymbolOrFallback(fallback: string): symbol | string {
+  return typeof Symbol === 'function' ? Symbol() : fallback;
+}
```
I think we're OK to just require `Symbol` to be supported now: https://caniuse.com/?search=symbol

MackinnonBuck committed:
Improved DOM syncing

Still lots under construction. I'm expecting many E2E tests to fail because Boot.Web.ts currently doesn't initialize components until the DOM gets updated after the initial page load.

MackinnonBuck committed:
Working boundary syncing, WASM TODO

MackinnonBuck commented on src/Components/Server/src/Circuits/CircuitHost.cs:
Changes to this file can be pretty much ignored. There's lots of copy/pasted and unnecessary code that can be cleaned up later. I might favor using JS interop over hub methods later anyway.

MackinnonBuck commented on src/Components/Server/src/Circuits/ServerComponentDeserializer.cs:
```diff
@@ -105,6 +105,18 @@ public bool TryDeserializeComponentDescriptorCollection(string serializedCompone
                 return false;
             }
 
+            // We force the client to send descriptors in the order that they appear on the page
+            // as a way to ensure that descriptors don't get duplicated.
+            // The client might be lying about the sequence, but we validate its value
+            // after we unprotect the descriptor. This check is done early to avoid
+            // unprotecting descriptors that we know are invalid.
+            if (marker.Sequence <= lastSequence)
+            {
+                Log.InvalidMarkerSequence(_logger, lastSequence, marker.Sequence);
+                descriptors.Clear();
+                return false;
+            }
```
These changes are because we can't force the sequence to start at 0 anymore.

MackinnonBuck commented on src/Components/Server/src/ComponentHub.cs:
Ignore changes to this file too.

MackinnonBuck commented on src/Components/Server/test/Circuits/ServerComponentDeserializerTest.cs:
```diff
@@ -8,6 +8,11 @@
 
 namespace Microsoft.AspNetCore.Components.Server.Circuits;
 
+// TODO:
+// - Test that the marker sequence is required.
+// - Test that the marker sequence must match the descriptor sequence.
+// - Test that markers cannot be duplicated.
```
I will implement these after settling on the approach for registering dynamicaly-SSR'd components with .NET.

MackinnonBuck commented on src/Components/Web.JS/src/Boot.Server.Common.ts:
```diff
@@ -113,7 +126,7 @@ async function initializeConnection(options: CircuitStartOptions, logger: Logger
 
   const newConnection = connectionBuilder.build();
 
-  newConnection.on('JS.AttachComponent', (componentId, selector) => attachRootComponentToLogicalElement(0, circuit.resolveElement(selector), componentId, false));
```
I think this was a bug waiting to happen, and we only hadn't caught it yet because Blazor Server typically starts faster than Blazor WebAssembly. Thankfully this bug is only relevant to .NET 8.

I've changed the "renderer ID" assignment to use statically-assigned IDs to make this mechanism more robust.

MackinnonBuck commented on src/Components/Web.JS/src/Boot.Web.ts:
```diff
@@ -9,56 +9,121 @@
+      addServerDescriptor(descriptor);
+      break;
+    case 'webassembly':
+      startWebAssemblyIfNotStarted();
+      // TODO: WebAssembly.
+      break;
```
I haven't yet implemented the WebAssembly parts of this feature, but I believe those should be much simpler since interop can happen synchronously. This eliminates concerns like:
* What if an SSR component comment gets removed while the interactive component is getting added?
* What if the comment was replaced with a new one that has updated parameter values?
* etc.

MackinnonBuck commented on src/Components/Web.JS/src/Boot.Web.ts:
```diff
@@ -9,56 +9,121 @@
+  handleUpdatedDescriptors();
+}
+
+let webAssemblyStarted = false;
+async function startWebAssemblyIfNotStarted() {
+  if (webAssemblyStarted) {
+    return;
+  }
+
+  webAssemblyStarted = true;
+  await startWebAssembly(webStartOptions?.webAssembly);
+  handleUpdatedDescriptors();
```
I know there's a lot of "Server this" and "WebAssembly that" in this file, but that's hopefully going to go away after I get the WebAssembly functionality working and know what the right abstractions are.

MackinnonBuck commented on src/Components/Web.JS/src/Rendering/BrowserRenderer.ts:
```diff
@@ -393,22 +393,15 @@ function clearElement(element: Element) {
 }
 
 function clearBetween(start: Node, end: Node): void {
-  const logicalParent = getLogicalParent(start as unknown as LogicalElement);
-  if (!logicalParent) {
-    throw new Error("Can't clear between nodes. The start node does not have a logical parent.");
-  }
-  const children = getLogicalChildrenArray(logicalParent);
-  const removeStart = children.indexOf(start as unknown as LogicalElement) + 1;
-  const endIndex = children.indexOf(end as unknown as LogicalElement);
-
-  // We remove the end component comment from the DOM as we don't need it after this point.
-  for (let i = removeStart; i <= endIndex; i++) {
-    removeLogicalChild(logicalParent, removeStart);
-  }
```
Related to a change in `LogicalElements.ts`

SteveSandersonMS replied:
I think the existing logic will be required. Unless we use `removeLogicalChild`, the parent's list of logical children will no longer be in sync with the DOM.

MackinnonBuck commented on src/Components/Web.JS/src/Rendering/LogicalElements.ts:
```diff
@@ -49,16 +53,21 @@ export function toLogicalRootCommentElement(start: Comment, end: Comment): Logic
-  const parent = start.parentNode;
   const parentLogicalElement = toLogicalElement(parent, /* allow existing contents */ true);
   const children = getLogicalChildrenArray(parentLogicalElement);
-  Array.from(parent.childNodes).forEach(n => children.push(n as unknown as LogicalElement));
```
This was causing a bug where if two or more components get prerendered inside the same DOM element, the logical children array gets duplicate nodes. This results in weird behavior where components might swap places in the DOM after being activated.

I might be missing some context, but I'm not sure why it was done this way instead of only treating the 'start' and 'end' comments as logical elements. The content between them will get deleted soon after this anyway. This change didn't cause any problems in my testing, but maybe others have insight into the original thinking behind this.

SteveSandersonMS replied:
> This was causing a bug where if two or more components get prerendered inside the same DOM element, the logical children array gets duplicate nodes.

I totally see why that would have happened - the older logic simply failed to account for this situation.

However I would expect the solution here to be simply checking whether the existing parent is already configured as a logical element, and if so, not rebuilding its array of children. If it's already a LogicalElement, then we just have to update the `start` to have `start[logicalEndSiblingPropname] = end` so that when we're walking the siblings we automatically skip over the content nodes between start and end.

> I might be missing some context, but I'm not sure why it was done this way instead of only treating the 'start' and 'end' comments as logical elements

We can't only treat the start/end as logical elements because then the parent would not have an accurate list of its children. If someone called the LogicalElements APIs to mutate the parent's set of children in some way, it would get confused because the logical view wouldn't be in sync with the physical reality.

To keep things consistent, if the parent is being configured as a logical element, it needs an accurate list of children, which includes child nodes that are LogicalElements and ones that are not.

MackinnonBuck commented on src/Components/Web/src/WebRenderer.cs:
```diff
@@ -140,16 +133,20 @@ public Task DispatchEventAsync(JsonElement eventDescriptor, JsonElement eventArg
+        [JSInvokable]
+        public void RemoveRootComponent(int componentId)
+            => _renderer.RemoveRootComponent(componentId);
+
         [JSInvokable] // Linker preserves this if you call RootComponents.Add
-        public int AddRootComponent(string identifier, string domElementSelector)
+        public int AddJSRootComponent(string identifier, string domElementSelector)
             => _jsComponentInterop.AddRootComponent(identifier, domElementSelector);
 
         [JSInvokable] // Linker preserves this if you call RootComponents.Add
-        public void SetRootComponentParameters(int componentId, int parameterCount, JsonElement parametersJson)
+        public void SetJSRootComponentParameters(int componentId, int parameterCount, JsonElement parametersJson)
             => _jsComponentInterop.SetRootComponentParameters(componentId, parameterCount, parametersJson, _jsonOptions);
 
         [JSInvokable] // Linker preserves this if you call RootComponents.Add
-        public void RemoveRootComponent(int componentId)
+        public void RemoveJSRootComponent(int componentId)
```
These changes aren't final, like many of the current changes on the .NET side. The C# changes will be more ready-to-review when I finish WebAssembly support for this feature.

MackinnonBuck committed:
Minor cleanup

MackinnonBuck commented:
Also TODO is writing thorough E2E tests that capture all the edge cases. I've updated the test app to demonstrate a mainstream scenario, but there are a lot of intricacies that still need validation.

MackinnonBuck committed:
WebAssembly support

MackinnonBuck commented on src/Components/Server/src/Circuits/ServerComponentDeserializer.cs:
```diff
@@ -147,6 +147,12 @@ public bool TryDeserializeComponentDescriptorCollection(string serializedCompone
         return true;
     }
 
+    public ComponentDescriptor DeserializeServerComponentDescriptor(ServerComponentMarker record)
+    {
+        var (descriptor, _) = DeserializeServerComponent(record);
+        return descriptor;
+    }
+
```
The requirement to deserialize descriptors in the order that they were rendered (and without skipping any) doesn't map perfectly to this new rendering model. Descriptors could be skipped due to being replaced or removed on the page before getting activated as interactive components. Furthermore, some situations warrant handling descriptors in the wrong order. For example, components added during a streaming update don't become interactive components until the streaming request completes. However, already-interactive components should receive parameter updates immediately.

Fundamentally, we are trying to make sure that descriptors are legitimate and can't get used more than once. Data protection should guarantee the legitimacy of the descriptor, so we would just need to find an alternative way to guarantee that they don't get reused.

SteveSandersonMS replied:
> we would just need to find an alternative way to guarantee that they don't get reused.

Would it be enough just to have a `HashSet<int>` on `RemoteRenderer` tracking the ones we've consumed? If we know DataProtection ensures the client can't make up new ID numbers on its own, then the total size would be bounded.

MackinnonBuck replied:
Something like that should work. The other thing to consider is that the sequence ID will start over at 0 for each new HTTP response, so we'll have to pair it with the invocation ID (which is unique to the HTTP response) as well.

If we're worried about the hash set becoming too large, we could make this slightly better by adding a rule where after we deserialize a descriptor with an unseen invocation ID, all previously seen invocation IDs become invalid. So, you have two hash sets, one that stores consumed invocation IDs, and one that stores consumed sequence IDs for the current invocation ID.

MackinnonBuck committed:
Update EndpointHtmlRenderer.Streaming.cs

SteveSandersonMS commented on src/Components/Server/src/Circuits/RemoteRenderer.cs:
```diff
@@ -59,12 +64,75 @@ public Task AddComponentAsync(Type componentType, ParameterView parameters, stri
     }
 
+    // This ID needs to be kept in sync with the Blazor JS implementation.
+    protected override int AllocateRendererId() => 1;
```
Would it make sense to call this `GetWebRendererId`? Using the term "allocate" suggests distinct values would be returned.

SteveSandersonMS replied:
Also, perhaps nitpicky, but there could be a shared-source enum giving these IDs so it will be more obvious to us in the future where they are centrally organized and we won't have to worry about how to be sure the values are unique.

MackinnonBuck replied:
I agree, `GetWebRendererId` sounds better overall and is more descriptive of what's actually happening.

I think the reason I decided on the "allocate" terminology is that there's already a `RendererId` property that has a getter. I know not many people are going to be actually interacting with the renderer directly, but I thought having both `RendererId` and `Get{Web}RendererId()` could be confusing (i.e., which one should I use if I'm trying to get the renderer ID?). The "allocate" term makes it a little scarier in the list of intellisense options. But that name does seem misleading.

Maybe `AssignRendererId()`? But that's also not accurate.

Agh, I'm overthinking it. `GetWebRendererId()` sound fine to me 🙂

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.Web.ts:
```diff
@@ -15,50 +15,113 @@ import { shouldAutoStart } from './BootCommon';
+import { RendererId } from './Rendering/RendererId';
 
 let started = false;
+let isPerformingEnhnacedNavigation = false;
```
```suggestion
let isPerformingEnhancedNavigation = false;
```

MackinnonBuck replied:
I'm amazed that you spotted this. Even with the change highlighted it took me 10 seconds to figure out what was wrong.

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.Web.ts:
```diff
@@ -15,50 +15,113 @@ import { shouldAutoStart } from './BootCommon';
-    // This is because, otherwise, we would need a way to add new interactive root components to an existing circuit and that's #48763
-    detachProgressivelyEnhancedNavigationListener();
+function beforeDocumentUpdated() {
+  lastActiveElement = document.activeElement;
```
Hopefully this won't be required once we start leaving the interactive-component DOM in the document while merging in the new static content.

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.WebAssembly.Common.ts:
```diff
@@ -16,8 +16,9 @@ import { JSInitializer } from './JSInitializers/JSInitializers';
+import { RootComponentManager } from './Services/RootComponentManager';
 
-export async function startWebAssembly(options?: Partial<WebAssemblyStartOptions>, components?: WebAssemblyComponentDescriptor[]): Promise<void> {
+export async function startWebAssembly(options?: Partial<WebAssemblyStartOptions>, components?: WebAssemblyComponentDescriptor[] | RootComponentManager): Promise<void> {
```
Would it be practical to switch over to use `RootComponentManager` in all cases? Also on `CircuitManager.ts`.

MackinnonBuck replied:
Possibly - the only downside I can think of is that when using `blazor.server.js` or `blazor.webassembly.js`, there would be an added interop call after app startup to add root components, rather than having them be added instantly on startup. I'm not sure if this difference in behavior is meaningful, but I thought I'd play it safe to minimize how much gets changed in this PR.

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/DomMerging/DomSync.ts:
```diff
@@ -43,26 +52,34 @@ export function synchronizeDomContent(destination: CommentBoundedRange | Node, n
+          const source = nextNewContentNode!;
           nextDestinationNode = nextDestinationNode!.nextSibling;
           nextNewContentNode = nextNewContentNode!.nextSibling;
+          treatAsMatch(destination, source);
```
These sorts of changes could be eliminated if we had an abstraction over the DOM APIs here that presented a view of the world that incorporates "startmarker ... contents ... endmarker" as a single logical node. I *think* that's something you mentioned you'd be considering doing when we chatted.

SteveSandersonMS replied:
The existing `LogicalElements` APIs *might* be a sufficient representation, as they already have the concept that marker comments can be used to represent a single logical node with children. We just need to take the static content, find all the marker nodes in them, and upgrade them to be LogicalElements in the same way that `toLogicalRootCommentElement` already does. Then the logic in `DomSync` would need to use the `LogicalElements` APIs when walking the DOM so it sees it with respect to that view of the world.

Part of the reason why `synchronizeDomContent` accepts a `CommentBoundedRange` as a way to define the parent node is for precisely this case - we may want to treat a logical element as being the parent.

MackinnonBuck committed:
Don't remove components from DOM

MackinnonBuck committed:
Support for '@key', other improvements

MackinnonBuck committed:
Merge remote-tracking branch 'origin/main' into mbuck/ssr-boundary-improvements

MackinnonBuck committed:
Small fixes

MackinnonBuck committed:
Create blazor.webview.js

MackinnonBuck committed:
Cleanup

MackinnonBuck reviewed (commented):
I think this PR is in a good place now. The big thing that I still need to do is write E2E tests, which is what I'll be focusing on over the next few days. That said, I've been validating these changes with some manual testing scenarios (one of which is included in this PR), so I'm fairly confident in the overall stability of this feature.

MackinnonBuck committed:
Undo small change

MackinnonBuck commented on src/Components/Web.JS/src/Rendering/LogicalElements.ts:
```diff
@@ -49,36 +51,71 @@ export function toLogicalRootCommentElement(start: Comment, end: Comment): Logic
-  // We might not have an end comment in the case of non-prerendered components.
+  start[logicalRootDescriptorPropname] = descriptor;
+  const startLogicalElement = toLogicalElement(start);
+
   if (end) {
-    start[logicalEndSiblingPropname] = end;
-    toLogicalElement(end);
+    // We need to make each element between the start and end comments a logical child
+    // of the start node.
+    const rootCommentChildren = getLogicalChildrenArray(startLogicalElement);
+    const startNextChildIndex = Array.prototype.indexOf.call(children, startLogicalElement) + 1;
+    let lastMovedChild: LogicalElement | null = null;
+
+    while (lastMovedChild !== end as unknown as LogicalElement) {
+      const childToMove = children.splice(startNextChildIndex, 1)[0];
+      if (!childToMove) {
+        throw new Error('Could not find the end component comment in the parent logical node list');
+      }
+      childToMove[logicalParentPropname] = start;
+      rootCommentChildren.push(childToMove);
+      lastMovedChild = childToMove;
+    }
```
The updated DOM sync algorithm heavily relies on a sound logical element hierarchy, even for prerendered content, hence these changes. This wasn't really a requirement in the past because the only operation we did on prerendered content was remove it, and we had a special "logical end sibling" property to help with that case.

SteveSandersonMS replied:
The sheer amount of sophistication and complexity in this is a bit of an alarm. Perhaps I'm missing key scenarios but all I thought we needed to do was:

 * During DOM sync, when we're walking through a list of sibling elements, if we ever see an interactive component marker:
   * Treat that as a distinct new element type with its own separate logic for insert/update/delete (i.e., insertions/updates/deletions are operations on the corresponding interactive root component)
   * Then continue the walk by skipping to the node after the "end" marker, if there is one

I suggested that LogicalElement might be a helpful abstraction for seeing the DOM in a way that already has helpers for getting "next sibling" etc with respect to this virtual hierarchy. However we should only use that if it really is helpful and convenient.

The impression I get here is that interactive rendering's use of LogicalElement has pretty different requirements to the usage during DomSync. As such, trying to unify this into an all-encompassing system might be forcing a lot of complexity and perhaps would compromise on perf too (e.g., needing this new rule that "if a node is a LogicalElement, so are all its descendants").

I appreciate there isn't a lot of free time to finish this off! However we do have to keep maintaining `DomSync.ts` and `LogicalElement.ts` in the long run. TBH `LogicalElement` already was scary to me before this change because it's really intricate and has so many special cases that I can't remember, and I just trust that it's so battle-hardened by now that all these special cases do actually together in any combination that can actually occur. There would be a good case for not expanding it to accommodate DomSync if DomSync's requirements can actually be met with a simpler alternative.

What do you think?

MackinnonBuck committed:
Small fixes

MackinnonBuck committed:
Update SSRRenderModeBoundary.cs

SteveSandersonMS commented on src/Components/Server/src/Circuits/RemoteRenderer.cs:
```diff
@@ -59,12 +64,90 @@ public Task AddComponentAsync(Type componentType, ParameterView parameters, stri
+                return;
+            }
+
+            _ = RenderRootComponentAsync(componentId, descriptor.Parameters);
```
Just to be sure, we've definitely been through DataProtection checks before we get here, right? We definitely need not to let clients supply arbitrary parameters or arbitrary component IDs.

MackinnonBuck replied:
Yes, that happens in the call to `TryDeserializeSingleComponentDescriptor`.

MackinnonBuck commented on src/Components/Server/src/Circuits/RemoteRenderer.cs:
```diff
@@ -59,12 +64,90 @@ public Task AddComponentAsync(Type componentType, ParameterView parameters, stri
+            }
+        }
+
+        return;
```
This was intentional to make it clear that there isn't some code hiding after the local function declarations. I can remove it if you feel that it adds more confusion than it eliminates!

javiercn replied:
Maybe you can consider not putting these local functions if their "combined size" is so big that makes you lose track of what's going on in the method. They might be better off as separate members.

MackinnonBuck commented on src/Components/Server/src/Circuits/ServerComponentDeserializer.cs:
```diff
@@ -147,6 +148,75 @@ public bool TryDeserializeComponentDescriptorCollection(string serializedCompone
+
+        // We're seeing a different invocation ID than we did the last time we processed a component.
+        // There are two possibilities:
+        // [1] A new invocation has started, in which case we should stop accepting components from the previous one.
```
Yes, it's based on all of those expectations. For that third one, I couldn't think of a scenario that would contradict that assumption.

If we're worried about these expectations not holding up, we could take an alternate approach where we store a hash set of some invocation ID/sequence pair, and we wouldn't have to make any of the above assumptions. The current assumptions only allow us to stop tracking information about SSR renders that we know aren't supposed to be relevant anymore.

javiercn commented on src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs:
```diff
@@ -106,15 +108,20 @@ private void Prerender(RenderTreeBuilder builder)
+
+    private string GenerateMarkerKey(int sequence, object? key)
+    {
+        return HashCode.Combine(_componentType.FullName, sequence, key).ToString("x", CultureInfo.InvariantCulture);
```
Have we measured the chances of collision here? `GetHashCode` is not collision resistant.

MackinnonBuck replied:
I left a comment about that [here](https://github.com/dotnet/aspnetcore/pull/49238#discussion_r1264762420). I'm definitely open to improving this, because a collision is definitely possible.

We could reduce the possibility of a collision by generating a string that separately includes the following:
* Sequence number
* Key hash code
* Type name hash code

We could even go further and obtain a SHA1 hash for the type name.

javiercn replied:
I think we should still send the type name + sequence number without hashing. We can hash the key if there is any, but if we are using this for comparison, it feels like we are shooting ourselves in the foot computing a "digest" for the key when we have all the data.

If we are concerned about the size of the TypeName, we can create a Map for each type that we render as a server component on the fly that guarantees their uniqueness, but I would start with just sending the data without losing any information.

MackinnonBuck replied:
And is it fine that we might be sending the full name of a server-only type to the client? I wasn't sure if that would be considered a security risk. e.g., `Company.App.SomeComponent_AndByTheWayThePasswordIsFoo`.

javiercn replied:
It's inside a data protected payload, isn't it?

MackinnonBuck replied:
The key is not data protected. It wouldn't work to included it in the data protected payload because any change to parameter values would also change the payload's text. We could include it as a separate data protected payload, but then that's just a more expensive hash (and the key doesn't get round-tripped back to the server again anyway, it's only used by the JS code).

MackinnonBuck replied:
@SteveSandersonMS I agree that from a security perspective, it's probably fine to hash the key and other data alongside it.

What do you think of the concern about hash collisions? Do you think it would be fine to alternatively send the full name of the component type to the client, and possibly suggest/require that the key implements `IFormattable`?

SteveSandersonMS replied:
I was thinking about hash collisions a bit yesterday.

It's not as risky as it would be if we needed every hash to be unique - if that were the case then a list of only ~4500 items would have a > 50% chance of a clash (via birthday paradox maths). We only need the keys to be unique enough that the edit operations that happen in practice don't run into clashes. If the edits are deletions/insertions, then we only need the *next/preceding* item not to have the same hash as the inserted/deleted item. So the probability of this being violated is simply 1/2^32. If the edit operation was randomly shuffling a list, then the risk is harder to compute (I lost interest after a few minutes...).

> alternatively send the full name of the component type to the client

I don't think that helps much since the sequence number almost always uniquely identifies the type anyway. In fact it always would if the code was compiled from `.razor`, since there's no way in Razor syntax to render a child whose type varies without the sequence number varying. The one case where this is not true is if you deploy a new version of your app in between the two requests - that's the one edge case I think it's worth retaining a hash of the type name for.

> possibly suggest/require that the key implements IFormattable?

Yeah, that is a good possible solution. We could say that we will *only* include the `@key` value in the marker if it implements `IFormattable` (and if it doesn't, that's not an error since it may be used for interactive rendering, but we simply ignore the key in that case).

The other good thing about this is it avoids false negatives. That is, if someone was using `@key=someModelObjectFromTheDatabase` which is completely valid and useful in interactive rendering:

 * If we use `(type, sequence, key?.GetHashCode())`, it will produce a different output on every request, perhaps causing us *never* to retain the interactive components
 * If we only use `(type, sequence, (key as IFormattable)?.ToString())` then we will err on the side of retaining interactive components, and `@key` becomes a way to optionally prevent unwanted retention in a predictable way. It also guarantees no false positive matches.

So, your `IFormattable` suggestion sounds like a better solution to me.

SteveSandersonMS replied:
> The one case where this is not true is if you deploy a new version of your app in between the two requests - that's the one edge case I think it's worth retaining a hash of the type name for.

Actually forget that - it's an irrelevant scenario. If there is a new version of the app between the two requests, it would only affect WebAssembly because that's the only kind of interactive component that could survive a server restart.

It *is* still relevant to avoid accidental matches based on `sequence` alone because the circuit would reject any incoming parameters for a componentId that doesn't match the component type encoded in the Data Protected payload[1]. So there'd be an exception if there's an accidental match for a different component type. And that could happen because when we merge the DOM trees, some parent component might have contained an `@if`/`@else` branch that outputs two identical DOM structures that place two different component types at the equivalent output location and happen to have the same `sequence` number at that point in rendering.

So, to make things much more robust in the default case where there's no `@key`, I still think we should be including some information based on the component type. We could use the full type name if we are OK with disclosing that. Or if not, we'd want some stable value derived from it. But then we have to decide if we rely on SSR client-server affinity or not. If we want to allow for a cluster of SSR servers without client affinity, we cannot use either:

 * A `Type`->number lookup table maintained in server memory
 * `type.FullName.GetHashCode()`, since `GetHashCode` isn't required to be stable across processes

Instead we'd need a stable, server-independent hash of the type name. For example the server could maintain a cache of `Type`->`SHA(type.FullName)`.

[1] We do verify and enforce that, right?

MackinnonBuck replied:
Great analysis, @SteveSandersonMS!

> It is still relevant to avoid accidental matches based on sequence alone because the circuit would reject any incoming parameters for a `componentId` that doesn't match the component type encoded in the Data Protected payload... We do verify and enforce that, right?

Yes, if there is a type mismatch between the type in the data protected payload and the type of the component with the supplied ID, we reject the parameter update (see line 130 below).

https://github.com/dotnet/aspnetcore/blob/d3397906f11eb82a3b01780272f77d66eaf56765/src/Components/Server/src/Circuits/RemoteRenderer.cs#L123-L136

However, it's totally possible for the client to lie about the component ID if it happens to correspond to another root component of the same type. It would be tricky to come up with a mechanism to prevent that from happening, because it's the client who decides which SSR/interactive components match up anyway.

> `type.FullName.GetHashCode()`, since `GetHashCode` isn't required to be stable across processes

Interesting... I didn't know that was the case for strings. Good to know!

> For example the server could maintain a cache of Type->SHA(type.FullName)

This seems like the right approach to me.

Based on this discussion, it's looking like we'll generate a key with:
* The sequence number included directly
* A cached SHA for the component type name (maybe a truncated hash if we want to reduce payload size, as long as it doesn't make the risk of collision unacceptable)
* The formatted `@key`, if it's provided and implements `IFormattable`

javiercn commented on src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs:
```diff
@@ -92,11 +93,12 @@ private void Prerender(RenderTreeBuilder builder)
         var parameters = _latestParameters is null
             ? ParameterView.Empty
             : ParameterView.FromDictionary((IDictionary<string, object?>)_latestParameters);
+        var markerKey = GenerateMarkerKey(sequence, key);
```
Is the idea here that `@key` is "best effort"? Meaning:
1) It's not required.
2) It might not produce a unique value.
3) It might not produce a stable version between renders.

Or do we need it to produce a stable identifier for it to work. If not, what is the outcome?

MackinnonBuck replied:
> It's not required.

Correct. If it's not supplied, the sequence number and component type will be used to distinguish between component "instances".

> It might not produce a unique value... It might not produce a stable version between renders.

Right, if a `@key` is provided, it needs to produce a stable identifier. Maybe we could require that they `@key` implement `IFormattable`, and we use the formatted output in the key we generate for the marker. And if the `@key` happens to not implement `IFormattable`, we fall back on the hash code and log a warning. Worst case, the unstable key causes the component to get reinstantiated, but it shouldn't break app functionality in general. Or, alternatively, we could throw an exception to minimize the risk of shipping an app with this mistake.

SteveSandersonMS replied:
It's more than just the `@key` being best-effort. The *entire* matching and instance retention mechanism is best-effort. There can be both:

 * False negatives, where we don't retain an interactive instance because of a `@key` not being stable across renders/servers
 * False positives, where we do retain an interactive instance and supply parameters that the developer might think would be for a new instance, because of there being no `@key` or it clashing

In both cases this has to be *not* a correctness or security issue. How I think we should think of it:

* Each time there's an SSR render, the SSR server is signing some tokens that each give permission to have an interactive component of a particular type with a particular set of parameters.
* The client maintains a *separate* relationship with an interactive component server (might not be the same as the SSR server). Within this circuit, there are a set of interactive root components, and has a ComponentId for each of them. Whenever the client receives a token from an SSR server, the client is free choose either to:
  * Start a new interactive root component using this token, getting back a new ComponentId
  * Or, update *any* existing ComponentId in the circuit to use the parameters from this token. The only restriction is that the token encodes the component type, and the server will only accept the update if the ComponentId's existing type matches what the token says.
... and each time a token is sent to the interactive server, that token is burned and cannot be replayed.

Besides that, there are no rules. Neither the SSR server nor the interactive server can say whether any given match is expected/correct or not. Only the client maintains the state to decide whether to match or not, and an untrusted client can supply any token to update any ComponentId. But that's not a problem because by definition it's allowed for any root component to receive new parameters as long as they are from one of these signed tokens (and the component type matches).

The only real problem we're trying to avoid is accidentally getting exceptions from well-behaved clients getting false-positive matches that are rejected by the interactive server.

Does this match your expectations?

MackinnonBuck committed:
Merge remote-tracking branch 'origin/main' into mbuck/ssr-boundary-improvements

MackinnonBuck committed:
Stability improvements

MackinnonBuck committed:
More stability improvements

MackinnonBuck commented on src/Components/test/testassets/Components.TestServer/RazorComponents/Pages/Auth/InteractiveAuthenticationState.razor:
```diff
@@ -13,8 +13,8 @@
 
-<a href="@Nav.GetUriWithQueryParameter("username", "YourUsername")">Log in</a> |
-<a href="@Nav.GetUriWithQueryParameter("username", "")">Log out</a>
+<a href="@Nav.GetUriWithQueryParameter("username", "YourUsername")" onclick="location.reload()">Log in</a> |
+<a href="@Nav.GetUriWithQueryParameter("username", "")" onclick="location.reload()">Log out</a>
```
@SteveSandersonMS FYI, this change was required in order to make the `CanUseServerAuthenticationState_Interactive` test pass.

Before the changes in this PR, we would refresh the page instead of performing an enhanced navigation if there were interactive components on the page. Since an enhanced navigation is now performed, there isn't an opportunity for an updated `Task<AuthenticationState>` to be supplied to the `CascadingAuthenticationStateConsumer` component (because the cascading value can't cross the render mode boundary).

SteveSandersonMS replied:
Makes sense. This does seem like an actual scenario that could occur in real apps too (logic inside an SSR render might change the user's authentication state, but the circuit won't update its auth state since that's entirely separate).

I guess that:

 * For auth state changes like "log in" and "log out", you would either have to cause that to happen inside the circuit, or you'd have to forcibly reload the page like your updated test logic does here
 * For auth state changes like "user roles have changed", you would have to do the same as above if you need the change to take effect instantly, or you could wait for something like `RevalidatingServerAuthenticationStateProvider` to detect the stale state if it's OK for it to take a few minutes to update.

In any case, your change to the test logic here looks fine to me!

MackinnonBuck committed:
Update InteractiveStreamingRenderingComponent.razor

MackinnonBuck committed:
Basic tests, lots more on the way

MackinnonBuck committed:
Update TestCircuitHost.cs

MackinnonBuck committed:
More tests

MackinnonBuck committed:
Improved marker key generation

MackinnonBuck commented:
I've added quite a few tests now and updated the marker key generation per the discussion in this PR.

SteveSandersonMS commented on src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs:
```diff
@@ -106,15 +116,52 @@ private void Prerender(RenderTreeBuilder builder)
+        if (!s_componentTypeNameHashCache.TryGetValue(componentType, out var hash))
+        {
+            hash = ComputeComponentTypeNameHash(componentType);
+            s_componentTypeNameHashCache.Add(componentType, hash);
```
Writes to a `Dictionary<T,U>` aren't threadsafe according to the [docs](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-8.0#thread-safety) so should this be a `ConcurrentDictionary`? Or do you know that this particular TryGetValue+Add pattern would be threadsafe?

SteveSandersonMS commented on src/Components/Server/test/Circuits/RemoteRendererTest.cs:
```diff
@@ -419,6 +426,239 @@ public async Task ThrowsIfWeReceivedAnAcknowledgeForANeverProducedBatch()
+
+        // Assert
+        await component.WaitForDisposeAsync().WaitAsync(Timeout); // Will timeout and throw if not disposed
+    }
```
Would it also work to add a test here showing the server won't let you add *too many* root components without disposing old ones? Not sure if that enforcement happens within the relevant layer for this unit test though.

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/StreamingRendering.ts:
```diff
@@ -64,20 +67,23 @@ class BlazorStreamingUpdate extends HTMLElement {
       }
     }
+
+    navigationEnhancementCallbacks.documentUpdated();
```
This doesn't seem ideal for performance (it's rescanning the entire document after each `<blazor-ssr>` element is added), and I'm not sure it's right for correctness either.

Don't we need to wait until streaming SSR is entirely finished before we start activating interactive components? The prerendered content may include ComponentState generated during an async loading process. The developer won't want interactive components to light up until they have that state, as the whole point of transferring this state is so that during interactive rendering the component doesn't show up in a loading state or duplicate the cost of loading the same data again.

Apologies if I'm misinterpreting what's going on.

MackinnonBuck replied:
Yeah, I might not have done a good job picking a descriptive name for that callback. It actually doesn't re-scan the document for `<blazor-ssr>` elements; its purpose is to notify the `RootComponentManager` (or whoever else is interested, in theory) that the document has changed. In the case of `RootComponentManager`, this callback causes it to check each of the component descriptors it already tracks to see if they have been updated or removed (and, if a stream render is not underway, create interactive components for each descriptor that hasn't yet been handled).

> Don't we need to wait until streaming SSR is entirely finished before we start activating interactive components?

Yes, and the current implementation does that. But I do think it would be useful to update parameters for (or remove) already-interactive components. The callback is just a way to signal to whoever's interested that the document has changed and some root-components related operations might need to occur. The actual scanning for component comments happens in the DOM sync algorithm for incoming content, so by the time the content gets sync'd with the document, the `RootComponentManager` is already aware of any new SSR'd components.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationEnhancement.ts:
```diff
@@ -142,6 +156,7 @@ export async function performEnhancedPageLoad(internalDestinationHref: string, f
         } else {
           // For non-get requests, we can't safely re-request, so just treat it as an error
           replaceDocumentWithPlainText(`Error: ${fetchOptions.method} request to ${internalDestinationHref} returned non-HTML content of type ${responseContentType || 'unspecified'}.`);
+          navigationEnhancementCallbacks.documentUpdated();
```
I have to admit I'm not really a fan of having *many* different calls to `documentUpdated` from lots of different points in the flow. It looks pretty fragile, like we might either add it somewhere it shouldn't go, or forget to add it in some new place, and then the behavior will be unpredictable.

Ideally we would only need to activate interactive components in a fairly constrained set of places:

 * When the initial document load or a subsequent enhanced navigation is entirely completed, but *only* if there's no ongoing streaming rendering and no other enhanced navigation has started
 * When some streaming rendering completes (as in, entirely completes, not just a single chunk of it)
 * (Potentially) when the WebAssembly runtime completes its async loading process, if that was triggered by having seen a webassembly component marker in one of the two cases above

I don't think we have to deal with cases that are in the middle of a process, since we don't want to activate interactive components until streaming SSR has entirely stopped (because if we did, we wouldn't be able to wait for ComponentState).

I don't think we have to deal with these completely fatal error cases (`replaceDocumentWithPlainText`) either, since there's no going back from these - the page is permanently dead and the user will just have to leave it, which will dispose everything.

What do you think? I appreciate I might be missing some subtleties about why it's been done this way.

SteveSandersonMS replied:
If the scenarios listed above are accurate, then instead of passing around a collection of callbacks, we could define some concept like a "SSRLoadingState" whereby:

 * Code can call `SSRLoadingState.addWorkItem(somePromise)` to represent a thing that's happening, ideally at the top level of an async process like "perform enhanced nav including streaming SSR"
 * `Boot.Web.ts` or whenever can register a callback like `SSRLoadingState.onLoadingStateChanged(callback)` that fires when we move from 0 pending work items to >= 1, or when we move from >= 1 to 0.

Feel free to ignore this if it's irrelevant. I was just trying to make a constructive suggestion but maybe I'm missing the reasons why it needs notifications at each step in an ongoing process.

SteveSandersonMS replied:
Update: @MackinnonBuck explained I was missing what's really going on here. `documentUpdated` does *not* scan the document nor activate any new interactive components (that only happens during DOM sync). Instead, `documentUpdated` notifies the root component manager that it may need to update the parameters on existing root components, since we do want those to be updatable at arbitrary points during streaming SSR. However it might be cleaner still to centralize this so that the calls to RootComponentManager happen during DOM sync (since at that point we already know precisely which marker is being updated), and then RootComponentManager can debounce these calls so it only results in a single message to the server. Then we don't need `documentUpdated` at all (or at least, not in these cases).

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/DomMerging/DomSync.ts:
```diff
@@ -1,38 +1,102 @@
+  synchronizeDomContentCore(destination, newContent);
+}
+
+function synchronizeDomContentCore(destination: CommentBoundedRange | Node, newContent: Node) {
```
Entirely mixing together the DOM sync algorithm with special-case knowledge of logical elements representation and the interactive component markers is not ideal and will make this pretty hard to understand and maintain in the future. DOM syncing is already fairly sophisticated so it would be extremely valuable to layer this in a way that the interactive component marker handling and logical element walking is handled externally to this code.

Is there a way of adding a bit more abstraction to `DomSync.ts` so that it doesn't need to know about interactive component markers and the difference between logical and physical elements? What you have with `EditWalker` is a good start - maybe the `synchronizeDomContent` export itself could be defined to receive some abstraction like that as the input and output data.

There is good test coverage for the existing `DomSync.ts` behavior in `src\Components\Web.JS\test\DomSync.test.ts`, so anything we add/change in `DomSync` should be reflected in unit tests. The ultimate ideal would be if it didn't really gain new behavior at all besides accepting a more general abstraction to represent the DOM structures it's working over!

What do you think? Or is it impractical to add an abstraction that's still able to deal with some necessary special cases?

SteveSandersonMS commented on src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs:
```diff
@@ -106,15 +109,24 @@ private void Prerender(RenderTreeBuilder builder)
+            frame.Sequence,
+            frame.ComponentKey)
+            .ToString("x", CultureInfo.InvariantCulture);
+    }
```
Resolving this thread as the implementation is different now.

SteveSandersonMS reviewed (approved):
We discussed verbally and decided that:

 * There may be some opportunities for refinement that @MackinnonBuck can do today (e.g., maybe centralizing the `documentUpdated` callback mechanism)
 * But we don't want any polish phase to take too long here, since we're on a deadline for preview 7. So I'll approve this now and you can merge whenever you feel it's ready, based on how much time you have available.
 * Later, we may explore other possible ways to decouple the DomSync logic from the internals of LogicalElement and component markers. I may look at this a bit tomorrow if I have time.

MackinnonBuck committed:
PR feedback

MackinnonBuck commented:
Thanks for all the helpful feedback, @SteveSandersonMS and @javiercn.

Some follow-up items for RC1:
* Review the strategy we use to ensure clients are well-behaved when adding interactive root components
* Determine if we can simplify the DomSync algorithm to not care about the existence of logical elements. Or, if it does have to, use different layering to separate out logic specific to component comments and logical elements.
